src/lib/mina_base/account_update.ml

```diff
@@ -150,12 +152,46 @@ module Call_type = struct
     | s ->
         failwithf "Invalid call type: %s" s ()
 
+  let is_delegate_call = function Call -> false | Delegate_call -> true
+
+  let from_delegate_call = function false -> Call | true -> Delegate_call
+
   let quickcheck_generator =
-    Quickcheck.Generator.map Bool.quickcheck_generator ~f:(function
-      | false ->
-          Call
-      | true ->
-          Delegate_call )
+    Quickcheck.Generator.map ~f:from_delegate_call Bool.quickcheck_generator
+
+  let deriver obj =
+    let open Fields_derivers_zkapps in
+    iso_string ~name:"CallType" ~js_type:(Custom "CallType") ~to_string
+      ~of_string obj
```

Comment:
Wouldn't this be simpler if it were represented as a boolean, to not introduce another type in JS? Also it would be consistent with the circuit type

Reply:
Seeing how the introduction of a new leaf type `Custom "CallType"` [complicates the JS implementation](https://github.com/o1-labs/snarkyjs/commit/a9d8a2722ba32ed9c6c10c6508ed6f0bb7b5365c), I really think we should change this so that the JS type ends up being composed of existing simple types. Since you introduce a third call type in the next PR, I'd make this a record `{ isDelegateCall: boolean }` where we can later add a second field `{ isDelegateCall: boolean; isBlind: boolean }` or similar.

The simplest option, IMO, is to use a record for both the graphQL and the JS representation. For graphQL, a record of booleans is almost self-documenting, while a custom graphQL scalar is not. Here's an example of using `iso_record` to represent an OCaml enum as a record in graphQL and JS: https://github.com/MinaProtocol/mina/blob/f2fbd2f0090649f3c19bd2160556c28472147a5a/src/lib/mina_base/control.ml#L135-L137

If you do want to keep the enum representation in graphQL, which I'm OK with, then there is a second option: annotate this so that the _(in-circuit) JS representation_ becomes a record of booleans. To annotate a JS type that's different from the JSON type, there is `Fields_derivers_zkapps.with_checked`. Here is an example of annotating the `SequenceState`, which is just a field element, with `with_checked`: https://github.com/MinaProtocol/mina/blob/f2fbd2f0090649f3c19bd2160556c28472147a5a/src/lib/mina_base/zkapp_precondition.ml#L542-L543 (the sequence state needs this annotation because it has a different dummy value than a normal field element)

The only thing we need to add in JS, if we go with the second option and use `with_checked`, is a custom toJSON / fromJSON implementation. If we go with the first option, there is no new type stuff we need to add on the JS side.